src/compat/chainrulescore.jl

```diff
@@ -1,11 +1,31 @@
 function ChainRulesCore.rrule(::typeof(getproperty), x::ComponentArray, s::Union{Symbol,Val})
-    function getproperty_adjoint(Δ)
-        zero_x = zero(similar(x, eltype(Δ)))
-        setproperty!(zero_x, s, Δ)
-        return (ChainRulesCore.NoTangent(), zero_x, ChainRulesCore.NoTangent())
-    end
+    return getproperty(x, s), Δ -> getproperty_adjoint(Δ, x, s)
+end
+
+function getproperty_adjoint(Δ, x, s)
+    zero_x = zero(similar(x, eltype(Δ)))
+    zero_x = __setproperty!(zero_x, s, Δ)
+    return (ChainRulesCore.NoTangent(), zero_x, ChainRulesCore.NoTangent())
+end
 
-    return getproperty(x, s), getproperty_adjoint
+__setproperty!(x, s, Δ) = __setproperty!(Val(false), x, s, Δ)
+function __setproperty!(::Val{false}, x, s, Δ)
+    setproperty!(x, s, Δ)
+    return x
+end
+# NOTE: I am not sure how this is avoiding the problem of mutation but if we wrap the
+#       mutating function into an `rrule` as done here, Zygote computes the correct
+#       gradient.
+__setproperty!(::Val{true}, x, s::Symbol, Δ) = __setproperty!(Val(true), x, Val(s), Δ)
```

Comment:
Is there anything we need to do (like `@inline` or `@constprop :agressive`) to make sure `s` constant propagates?

Reply:
> @constprop :none [ex]: disable constant propagation. This can reduce compile times for functions that Julia might otherwise deem worthy of constant-propagation. Common cases are for functions with Bool- or Symbol-valued arguments or keyword arguments.

From the docs. I am guessing Julia tries to constant propagate Symbols by default